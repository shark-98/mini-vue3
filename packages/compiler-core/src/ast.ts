import { CREATE_ELEMENT_VNODE } from "./runtimeHelpers";

export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION,
}
export const enum TagType {
  START,
  END,
}

// parse
export type contentType = string
export type contextType = {
  source: contentType
}
export interface nodeType {
  type: NodeTypes,
}
export interface interpolationType extends nodeType {
  content: nodeType & {
    content: contentType
  }
}
export type elementCodegenNodeType = {
  type: NodeTypes.ELEMENT,
  tag: elementType['tag'],
  props: any,
  children: any,
}
export interface elementType extends nodeType {
  tag: string,
  children: Array<elementType | interpolationType | textType>,
  codegenNode?: elementCodegenNodeType
}
export interface textType extends nodeType {
  content: contentType
}
export type childrenItemType = interpolationType | elementType | textType
export type childrenType = Array<childrenItemType>
export type rootType = { type: NodeTypes.ROOT, children: childrenType, codegenNode?: childrenItemType | elementType['codegenNode'], helpers?: any[] }
export type allChildrenType = rootType | childrenItemType

// transform
export type nodeTransformsType = Function[]
export type transformOptions = {
  nodeTransforms?: nodeTransformsType
}
export type transformContext = {
  root: rootType,
  helpers: Map<string, number>,
  helper(key: any): void
} & transformOptions

// generate
export type codegenContextType = {
  code: string,
  push: (source: string) => void
  helper(key: any): string
}

export const openDelimiter = '{{'
export const closeDelimiter = '}}'
export const startTag = '<'
export const closeTag = '>'
export const endTag = '</'

export function createVNodeCall(context: transformContext, tag: elementCodegenNodeType['tag'], props: elementCodegenNodeType['props'], children: elementCodegenNodeType['children']): elementCodegenNodeType {
  context.helper(CREATE_ELEMENT_VNODE);

  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children,
  };
}
