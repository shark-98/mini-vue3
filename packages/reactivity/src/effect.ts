import { extend } from "@mini-vue/shared"
import { anyObjectType, effectDepsType } from "@mini-vue/types"

let activeEffect: ReactiveEffect
let shouldTrack: boolean = false
const targetMap = new Map()

export class ReactiveEffect {
  private _fn
  deps: Array<any> = []
  active: boolean = true
  onStop?: () => void

  constructor(fn: Function, public scheduler?: Function) {
    this._fn = fn
  }

  run() {
    if (!this.active) {
      return this._fn()
    }

    // 应该收集
    shouldTrack = true
    activeEffect = this
    const res = this._fn()
    // 重置
    shouldTrack = false

    return res
  }

  stop() {
    if (this.active) {
      cleanupEffect(this)

      this.onStop?.()

      this.active = false
      shouldTrack = false
    }
  }
}

const cleanupEffect = (effect: ReactiveEffect) => {
  effect.deps.forEach((dep) => {
    dep.delete(effect)
  })

  // 把 effect.deps 清空
  effect.deps.length = 0
}

export const isTracking = () => {
  return shouldTrack && activeEffect !== undefined
}

export const track = (target: anyObjectType, key: string) => {
  if (!isTracking()) return

  let targetDep = targetMap.get(target)
  if (!targetDep) {
    targetDep = new Map()
    targetMap.set(target, targetDep)
  }

  let deps = targetDep.get(key)
  if (!deps) {
    deps = new Set()
    targetDep.set(key, deps)
  }

  trackEffect(deps)
}
export const trackEffect = (deps: effectDepsType) => {
  // 看看 dep 之前有没有添加过，添加过的话 那么就不添加了
  if (deps.has(activeEffect)) return

  deps.add(activeEffect)
  activeEffect.deps.push(deps)
}
export const trigger = (target: anyObjectType, key: string) => {
  const targetDep = targetMap.get(target)
  const deps = targetDep.get(key)
  triggerEffect(deps)
}
export const triggerEffect = (deps: effectDepsType) => {
  for (const dep of deps) {
    if (dep.scheduler) {
      dep.scheduler()
    } else {
      dep.run()
    }
  }
}

export const effect = (fn: Function, options: anyObjectType = {}) => {
  const _effect = new ReactiveEffect(fn, options.scheduler)
  extend(_effect, options)
  _effect.run()

  const runner: any = _effect.run.bind(_effect)
  runner._effect = _effect
  return runner
}
export const stop = (runner: any) => {
  runner._effect.stop()
}
