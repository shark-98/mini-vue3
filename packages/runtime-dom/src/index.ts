import { createRenderer } from '@mini-vue/runtime-core';

function createElement(type: any) {
  return document.createElement(type)
}

function patchProp(el: any, key: any, oldVal: any, newVal: any) {
  const isOn = (name: string) => /^on[A-Z]/.test(name)
  if (isOn(key)) {
    const event = key.slice(2).toLocaleLowerCase()
    el.addEventListener(event, newVal)
  } else {
    if ([undefined, null].includes(newVal)) {
      el.removeAttribute(key)
    } else {
      el.setAttribute(key, newVal)
    }
  }
}

function insert(child: any, parent: any, anchor: HTMLElement | null) {
  // parent.append(child)
  parent.insertBefore(child, anchor || null)
}

function remove(child: any) {
  const parent = child.parentNode
  if (parent) {
    parent.removeChild(child)
  }
}
function setElementText(el: HTMLElement, text: string) {
  el.textContent = text
}

const render: any = createRenderer({
  createElement,
  patchProp,
  insert,
  remove,
  setElementText
})

export function createApp(...args: any[]) {
  return render.createApp(...args)
}

export * from '@mini-vue/runtime-core'
