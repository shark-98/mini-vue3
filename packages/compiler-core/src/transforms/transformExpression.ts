import { allChildrenType, interpolationType, NodeTypes } from "../ast";

export function transformExpression(node: allChildrenType) {
  if (node.type === NodeTypes.INTERPOLATION) {
    (node as interpolationType).content = processExpression((node as interpolationType).content);
  }
}

function processExpression(node: any) {
  node.content = `_ctx.${node.content}`;
  return node;
}
