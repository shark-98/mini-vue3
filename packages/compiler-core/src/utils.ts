import { NodeTypes, allChildrenType } from './ast';

export function isText(node: allChildrenType) {
  return (
    node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION
  );
}
