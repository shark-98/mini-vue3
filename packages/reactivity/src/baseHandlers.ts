import { extend, isObject } from "@mini-vue/shared"
import { anyObjectType } from "@mini-vue/types"
import { track, trigger } from "./effect"
import { reactive, ReactiveFlags, readonly } from "./reactive"

function createGetter(isReadonly = false, shallow = false) {
  return (target: anyObjectType, key: string) => {
    if (key === ReactiveFlags.IS_REACTIVE) {
      return !isReadonly
    }
    if (key === ReactiveFlags.IS_READONLY) {
      return isReadonly
    }

    const res = Reflect.get(target, key)

    if (shallow) {
      return res
    }

    if (isObject(res)) {
      return isReadonly ? readonly(res) : reactive(res)
    }

    if (!isReadonly) {
      // 依赖收集track
      track(target, key)
    }

    return res
  }
}

function createSetter() {
  return (target: anyObjectType, key: string, value: any) => {
    const res = Reflect.set(target, key, value)

    // 依赖触发trigger
    trigger(target, key)
    return res
  }

}

const get = createGetter()
const set = createSetter()
const readonlyGet = createGetter(true)
const shallowReadonlyGet = createGetter(true, true)

export const reactiveHandler = {
  get,
  set
}
export const readonlyHandler = {
  get: readonlyGet,
  set(target: anyObjectType, key: string) {
    console.warn(`key: ${key} set Error, because target is readonly`)
    return true
  }
}
export const shallowReadonlyHandler = extend({}, readonlyHandler, { get: shallowReadonlyGet })
