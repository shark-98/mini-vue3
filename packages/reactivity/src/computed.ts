import { ReactiveEffect } from "./effect";

class ComputedRefImpl {
  private _dirty: Boolean = true;
  private _value: any;
  private _effect: ReactiveEffect;
  constructor(getter: Function) {
    this._effect = new ReactiveEffect(getter, () => {
      if (!this._dirty) {
        this._dirty = true
      }
    })
  }

  get value() {
    if (this._dirty) {
      this._dirty = false;
      this._value = this._effect.run()
    }

    return this._value
  }
}

export const computed = (getter: Function) => {
  return new ComputedRefImpl(getter)
}
