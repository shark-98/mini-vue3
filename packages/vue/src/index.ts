// mini-vue 出口
import { baseCompile } from '@mini-vue/compiler-core/src';
export * from '@mini-vue/runtime-dom'
import * as runtimeDom from "@mini-vue/runtime-dom";
import { registerRuntimeCompiler } from "@mini-vue/runtime-dom";

function compileToFunction(template: string) {
  const { code } = baseCompile(template);
  const render = new Function("Vue", code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compileToFunction);
