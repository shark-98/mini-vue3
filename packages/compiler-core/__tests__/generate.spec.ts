import { codegen } from '../src/codegen';
import { baseParse } from '../src/parse';
import { transform } from '../src/transform';
import { transformExpression } from '../src/transforms/transformExpression';
import { transformElement } from "../src/transforms/transformElement";
import { transformText } from "../src/transforms/transformText";

describe('generate', () => {
  it('string', () => {
    const ast = baseParse('hi')

    transform(ast)

    const { code } = codegen(ast);

    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}')

    transform(ast, {
      nodeTransforms: [transformExpression],
    })

    const { code } = codegen(ast);

    expect(code).toMatchSnapshot()
  })

  it('element', () => {
    const ast = baseParse('<div>hi,{{message}}</div>')

    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    })

    const { code } = codegen(ast);

    expect(code).toMatchSnapshot()
  })
})
