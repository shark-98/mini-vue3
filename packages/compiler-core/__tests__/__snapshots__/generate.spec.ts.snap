// Vitest Snapshot v1

exports[`generate > element 1`] = `
"const { toDisplayString:_toDisplayString, createElementVNode:_createElementVNode } = Vue
return function render(_ctx, _cache) {return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`generate > interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`generate > string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;

exports[`generate element 1`] = `
"const { toDisplayString:_toDisplayString, createElementVNode:_createElementVNode } = Vue
return function render(_ctx, _cache) {return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`generate interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`generate string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;
